fix(postgresql): properly handle definition of unique and PK constraints in ALTER TABLE statements

pierresouchay opened the pull request:
Will fix https://github.com/holistics/dbml/issues/548

## Summary
* Implemented parsing of ALTER TABLE ... ADD CONSTRAINT for "UNIQUE|PRIMARY KEY"

## Issue
https://github.com/holistics/dbml/issues/548

## Lasting Changes (Technical)

This adds the constraints as indexes (with their names). If 1 single column is impacted, this will also add the attribute of the impacted column of the table model.

## Checklist

Please check directly on the box once each of these are done

- [x] Documentation (if necessary)
- [x] Tests (integration test/unit test)
- [x] Integration Tests Passed
- [x] Code Review

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new table `comment_on_product` to enhance data structuring.
- **Enhancements**
	- Improved user data integrity by modifying the `email` field in the `users` table to be non-unique with a separate unique constraint.
	- Enhanced country identification by changing the `code` field in the `countries` table to have a dedicated primary key constraint.
- **Bug Fixes**
	- Fixed indexing issues by ensuring primary keys and unique constraints are correctly named and assigned.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pierresouchay commented on packages/dbml-core/src/parse/ANTLR/ASTGeneration/postgres/PostgresASTGen.js:
```diff
@@ -804,6 +804,37 @@ export default class PostgresASTGen extends PostgreSQLParserVisitor {
+              set_prop(index);
+              index.columns = cmd.value.columns;
+            }
+            break;
```
done

pierresouchay commented:
@NQPhuc I checked the changes in line by line in [packages/dbml-core/__tests__/importer/postgres_importer/output/db_dump.out.dbml](https://github.com/holistics/dbml/pull/549/files#diff-b1ef04086803b098fb6b1b12c89129f332b8e6865b0803c7d82a620531696ad7) => and everything looks fine (it adds lots of pk that were missing from initial test-case)

pierresouchay commented:
@NQPhuc I also had to fix to let pk have names in PyDBML: https://github.com/Vanderhoof/PyDBML/pull/35

=> Any comment on that? is
```dbml
Table{
  column1 int
  Index {
    column1 [pk, name: 'column1_idx']
  }
}
```
Is it a correct statement according to spec? For me, yes.

NQPhuc commented on packages/dbml-core/src/parse/ANTLR/ASTGeneration/postgres/PostgresASTGen.js:
```diff
@@ -804,6 +804,36 @@ export default class PostgresASTGen extends PostgreSQLParserVisitor {
             cmd.value.endpoints[0].schemaName = schemaName;
             this.data.refs.push(cmd.value);
             break;
+          case TABLE_CONSTRAINT_KIND.UNIQUE:
+          case TABLE_CONSTRAINT_KIND.INDEX:
+            kind = cmd.kind;
+            if (cmd.value.pk || cmd.value.unique) {
+              let set_prop = cmd.value.unique ? (obj) => obj["unique"] = true : (obj) => obj["pk"] = true;
+              const table = findTable(this.data.tables, schemaName, tableName);
+              if (!table){
+                break;
+              }
+              if (cmd.value.columns.length == 1) {
+                // We are in the single column case, we can update the column as well
+                let key = cmd.value.columns[0].value;
+                let field_to_update = table.fields.find(f => f.name === key);
+                // field might have been deleted
+                if (field_to_update != null) {
+                  set_prop(field_to_update);
+                }
+              }
+              let index = table.indexes.find(idx => { idx.name == cmd.value.name });
+              if (!index) {
+                index = new Index({
+                  name: cmd.value.name,
+                  columns: cmd.value.columns,
+                });
+                table.indexes.push(index);
+              }
+              set_prop(index);
+              index.columns = cmd.value.columns;
+            }
+            break;
```
Hi I suggest to update the logic to the below, the main points are:
* Generating both fields settings and indexes for the same constraint is not recommended, because when re-exported to SQL, it'll generate both a column constraint in CREATE TABLE column definition and a CREATE INDEX statement. Therefore. we should only choose one.
* IMO, using pk/unique field settings to describe SQL constraints is the better because it more inline with concept of SQL constraints, while indexes fit more with CREATE INDEX.. statements. However, it has two downsides:
  * It can only used for single column constraint -> I agree that for multi-column constraint, using to indexes is our only choice.
  * We can't name the pk/unique constraint in field settings -> For now, I recommend that for single-column named constraint, we still import it as a field property. In the future, we intend to add the ability to name field pk/unique property through this syntax in dbml. Please let me know what you think about this:
  ```
  table a {
    id int [pk: "pk_constraint_name"]
  }
  ...
  ```
```suggestion
          case TABLE_CONSTRAINT_KIND.UNIQUE:
          case TABLE_CONSTRAINT_KIND.INDEX: {
            if (!(cmd.value.pk || cmd.value.unique)) break;

            // eslint-disable-next-line prefer-destructuring
            kind = cmd.kind;

            const table = findTable(this.data.tables, schemaName, tableName);

            if (!table) break;

            // multi column constraint -> need to create new index
            if (cmd.value.columns.lenght > 1) {
              const index = new Index({
                name: cmd.value.name,
                columns: cmd.value.columns,
              });

              index[cmd.value.unique ? 'unique' : 'pk'] = true;

              table.indexes.push(index);

              break;
            }

            // single column contraint -> update the column instead
            const key = cmd.value.columns[0].value;
            const fieldToUpdate = table.fields.find(f => f.name === key);
            if (fieldToUpdate) fieldToUpdate[cmd.value.unique ? 'unique' : 'pk'] = true;

            break;
          }
```

pierresouchay replied:
@NQPhuc Ok, I am gonna take your path, I am gonna update this today (and update the test cases)

1. if constraint is on a single column, use "pk"/unique in column (without any name)
2. If constraint is on multiple columns, use "Indexes"

pierresouchay replied:
@NQPhuc  OK, done your way, this is ready!

NQPhuc replied:
Thanks, I think this PR is almost done. There's only one thing left: can you help add some test for ALTER TABLE ADD CONSTRAINT with multi-column to make sure it working properly?

NQPhuc commented:
> @NQPhuc I also had to fix to let pk have names in PyDBML: [Vanderhoof/PyDBML#35](https://github.com/Vanderhoof/PyDBML/pull/35)
> 
> => Any comment on that? Does it sound right to you?
> 
> ```
> Table my_table{
>   column1 integer
>   Index {
>     column1 [pk, name: 'column1_idx']
>   }
> }
> ```
> 
> Is it a correct statement according to spec? For me, yes.

@pierresouchay  Yes, for named pk indexes, this is correct according to spec. (minor thing but it should be `indexes` instead of `index`. But I believe that's just a typo)

pierresouchay commented:
> @pierresouchay Yes, for named pk indexes, this is correct according to spec. (minor thing but it should be `indexes` instead of `index`. But I believe that's just a typo)

Indeed, was a typo: Index should be Indexes

NQPhuc commented on packages/dbml-core/src/parse/ANTLR/ASTGeneration/postgres/PostgresASTGen.js:
```diff
@@ -804,6 +804,29 @@ export default class PostgresASTGen extends PostgreSQLParserVisitor {
             break;
+          case TABLE_CONSTRAINT_KIND.UNIQUE:
+          case TABLE_CONSTRAINT_KIND.INDEX:
+            if (cmd.value.columns == 0) break;
```
I think you mean columns.length? 🤔 
```suggestion
            if (cmd.value.columns.length === 0) break;
```

pierresouchay replied:
fixed

NQPhuc commented on packages/dbml-core/src/parse/ANTLR/ASTGeneration/postgres/PostgresASTGen.js:
```diff
@@ -804,6 +804,29 @@ export default class PostgresASTGen extends PostgreSQLParserVisitor {
+            if (!(cmd.value.pk || cmd.value.unique)) break;
+            // eslint-disable-next-line prefer-destructuring
+            kind = cmd.kind;
+            const table = findTable(this.data.tables, schemaName, tableName);
```
Declaring these without scoping them will give a lot of lint error, and potentially cause conflict with same name `const` in other cases. Please help enclosed the this whole code from  line 809 to line 829 within brackets { }. Thanks

pierresouchay commented:
@NQPhuc we introduced a regression in changes when not doing systematically indexes. Fixed that (typo).

Also added new tests to take into account multiple indexes. 

Also changed scoping of variables.

Should be good now
